client/src/components/ElementEditor/ElementList.js

```diff
@@ -11,6 +11,101 @@ import { getDragIndicatorIndex } from 'lib/dragHelpers';
 import { getElementTypeConfig } from 'state/editor/elementConfig';
 
 class ElementList extends Component {
+  constructor(props) {
+    super(props);
+    this.resetState = this.resetState.bind(this);
+    this.handleBeforeSubmitForm = this.handleBeforeSubmitForm.bind(this);
+    this.handleAfterSubmitResponse = this.handleAfterSubmitResponse.bind(this);
+    this.state = {
+      // saveAllElements will be set to true in entwine.js in the 'onbeforesubmitform' "hook"
+      // which is triggered by LeftAndMain submitForm()
+      saveAllElements: false,
+      // increment is also set in entwine.js in the 'onbeforesubmitform' "hook"
+      increment: 0,
+      hasUnsavedChangesBlockIDs: {},
+      validBlockIDs: {},
+    };
+    // Update the sharedObject so that setState() can be called from entwine.js
+    this.props.sharedObject.setState = this.setState.bind(this);
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    // Number of blocks just changed after graphql response
+    if (this.props.blocks !== prevProps.blocks) {
+      this.resetState(prevState, false);
+      return;
+    }
+    // Saving all elements and state has just updated because of a formSchema response from inline save
+    if (this.state.saveAllElements) {
+      const unsavedChangesBlockIDs = this.props.blocks
+        .map(block => parseInt(block.id, 10))
+        .filter(blockID => this.state.hasUnsavedChangesBlockIDs[blockID]);
+      const allValidated = unsavedChangesBlockIDs.every(blockID => this.state.validBlockIDs[blockID] !== null);
+      if (allValidated) {
+        const allValid = unsavedChangesBlockIDs.every(blockID => this.state.validBlockIDs[blockID]);
+        // entwineResolve is bound in entwine.js
+        this.props.sharedObject.entwineResolve(allValid);
+        this.resetState(prevState, allValid);
+        this.setState({ saveAllElements: false });
+      }
+    }
+  }
+
+  resetState(prevState, resetHasUnsavedChangesBlockIDs) {
+    if (this.props.blocks === null) {
+      return;
+    }
```

Comment:
Why do we not reset state if there's no blocks?

Reply:
Because we're looping this.props.blocks below, can't loop blocks if they're not there

Reply:
We should ensure `hasUnsavedChangesBlockIDs` and `validBlockIDs` are both empty though, surely?

Reply:
@emteknetnz Please respond to this

Reply:
As above - It's been too long since I've wrote this and I've lost context. If it's not working as expected when testing then we should change something, otherwise leave as is

Reply:
I don't know what edge cases might be caused and therefore don't know how to test them. I think it's worth the time to regain context on this and understand whether we need to fully reset the state here or not so we don't merge something that's broken in ways we don't understand.

Reply:
> We should ensure hasUnsavedChangesBlockIDs and validBlockIDs are both empty though, surely?

No, because we're resetting the component state.  Both of those variables are used in the call to setState below.

Reply:
The call to `setState()` below is skipped by this `return;` so that's not relevant. I'm talking about this condition, where there are no blocks and we are just returning.

We've called `resetState()`, but because there's no blocks, we return without doing anything? Seems like we should still reset these two parts of the state even when there's no blocks.

Reply:
It functionally makes no difference since they're already set to empty objects in the class constructor, though I'll make the update anyway just to make it a little clearer